### PR TITLE
Fix Recurring Events Handling 

### PR DESCRIPTION
**Pull request by abdullai-t**
####  Summary / Highlights

This pull request addresses issues with the handling of recurring events. Previously, recurring events were updated whenever a user reloaded the community portal site through an endpoint. That endpoint was removed when we decided to load data from the API on an as-needed basis; hence, the start and end dates of recurring events stopped updating.

This PR introduces one-off tasks for updating the start and end dates of recurring events. After updating these dates, a new task is scheduled to run at the current end date, continuing this process until the `final_date` is reached or the event is deleted.

### Todo After Merging

- Run `make create-init-schedule` to populate the database with initial tasks.
- Subsequent recurring event creations or updates will automatically create/update associated tasks.

#### Testing Steps (Provide details on how your changes can be tested)
run `make test`
#### Requirements (place an `x` in each `[ ]`)
* [x] I've read and understood the [Contributing Guidelines](/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've tested my changes manually.
* [ ] I've added unit tests to my PR.

##### Transparency ([Project board](https://github.com/orgs/massenergize/projects/7/views/2))
* [x] I've given my PR a meaningful title.
* [x] I linked this PR to the relevant issue.
* [ ] I moved the linked issue from the "Sprint backlog" to "In progress" when I started this.
* [ ] I moved the linked issue to "QA Verification" now that it is ready to merge.

### CODE REVIEW

File: src/api/tasks.py [hunk 1]
```diff
@@ -328,14 +331,59 @@ def automatically_activate_nudge(community_nudge_setting_id):
...
+    else:
+        log.info("Task already picked up by another worker")
+        
+        
+@app.task
+def update_recurring_event(event_id):
+    
+    cache_key = f"update_recurring_event_{event_id}"
+    
+    if cache.add(cache_key, True, timeout=3600):
```
Comment (Opoku-Agyemang) [inline src/api/tasks.py:351]: i dont think this is the right way to use a cache.
I think we should create a task with a name = the event's id or something and leverage the transaction logic we have for processing that one off task that runs at some future time = the event's end datetime

**Review by Opoku-Agyemang — COMMENTED** [review]
Looks good in general, left a note.  also please make sure the tests are passing

File: src/api/utils/api_utils.py [hunk 2]
```diff
@@ -327,3 +335,82 @@ def get_supported_language(language_code):
...
+
+# ------------------------ Recurring Events utils ----------------------------------------#
+
+def get_final_date(event, today):
+    start_time = event.start_date_and_time.strftime("%H:%M:%S+00:00")
+    final_date_str = event.recurring_details.get('final_date')
+    
+    if final_date_str and final_date_str != 'None':
+        final_date = datetime.strptime(f"{final_date_str} {start_time}", "%Y-%m-%d %H:%M:%S+00:00")
+        return final_date.replace(tzinfo=custom_timezone_info())
```
Comment (Opoku-Agyemang) [inline src/api/utils/api_utils.py:348]: nit: in general i think these helper functions can actually live in the events file themselves.  except that you want to use it in different places?

File: src/task_queue/database_tasks/update_recurring_events.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        return False
+
+
+def create_task_for_event(event, key):
+    recurring_details = get_recurring_details_from_date(event.end_date_and_time)
+    recurring_details_str = json.dumps(recurring_details) if recurring_details else None
+    task = Task(
+        name=key,
+        job_name="Update Recurring Events",
+        status="PENDING",
```
Comment (Opoku-Agyemang) [inline src/task_queue/database_tasks/update_recurring_events.py:53]: nit: things like PENDING, ONE_OFF should be defined as constants

**Review by Opoku-Agyemang — APPROVED** [review]

File: src/api/utils/api_utils.py (hunk 2 above)
Comment (abdullai-t) [reply src/api/utils/api_utils.py:348]: For now, the functions are used in only one place. I will move them to the events file.

File: src/task_queue/database_tasks/update_recurring_events.py (hunk 3 above)
Comment (abdullai-t) [reply src/task_queue/database_tasks/update_recurring_events.py:53]: got it 👍🏿

**Pull request merged**